k2/csrc/cuda/log.h

```diff
@@ -0,0 +1,119 @@
+// k2/csrc/cuda/log.h
+
+// Copyright (c)  2020  Fangjun Kuang (csukuangfj@gmail.com)
+
+// See ../../../LICENSE for clarification regarding multiple authors
+
+#ifndef K2_CSRC_CUDA_LOG_H_
+#define K2_CSRC_CUDA_LOG_H_
+
+#include <stdio.h>
+
+#include <cstdint>
+
+namespace k2 {
+
+namespace internal {
+
+enum class LogLevel {
+  kDebug = 0,
+  kInfo = 1,
+  kWarning = 2,
+  kError = 3,
+  kFatal = 4,
+};
+
+// They are used in LOG(xxx), so their names
+// do not follow the google c++ code style
+constexpr LogLevel DEBUG = LogLevel::kDebug;
+constexpr LogLevel INFO = LogLevel::kInfo;
+constexpr LogLevel WARNING = LogLevel::kWarning;
+constexpr LogLevel ERROR = LogLevel::kError;
+constexpr LogLevel FATAL = LogLevel::kFatal;
+
+class Logger {
+ public:
+  explicit __host__ __device__ Logger(LogLevel level) : level_(level) {}
+
+  __host__ __device__ ~Logger() {
+    printf("\n");
+    if (level_ == FATAL) {
+#if defined(__CUDA_ARCH__)
+      // abort() is not available for device code.
+      __threadfence();
+      asm("trap;");
+#else
+      abort();
+#endif
+    }
+  }
+
+  __host__ __device__ const Logger &operator<<(const char *s) const {
```

Comment:
@songmeixu  I think it's OK.  Full C++ is supported on the device, including c++14, it's just the standard C library
which is not fully supported, so the operator itself is not the issue.

I am still open to using glog but you need to explain to us your previous comments, better.

Reply:
Please check my below comments @danpovey≈

Reply:
I know why it's not working now. As the printf that rewritten by nvidia must have some thread guard for the stream. And that's also the reason why we shouldn't log things in cuda, it's performance killer. Just print msg if error happens is OK, that's why I put `K2_MAKE_ERROR_CHECK` to silence some msg from cuda in debug.h, @danpovey.

Reply:
> IDK much about cuda, but does it work on host&device to overload "<<" like this? 

I have never opened a pullrequest without making my code pass the local unit tests.

Reply:
The unit test would say nothing for right case..

Reply:
I cannot see the "<<" problem is solved.

Reply:
> The unit test would say nothing for right case..

I've added some death tests. You can see what the error messages look like from the logs in this [colab notebook][1]

[1]: https://colab.research.google.com/drive/1qbHUhNZUX7AYEpqnZyf29Lrz2IPHBGlX?usp=sharing#scrollTo=nk2rnCsB17Wa

Reply:
> I cannot see the "<<" problem is solved.

Normally please avoid printing messages in the kernel. The race condition happens
**only** in case of a check error with multiple threads without any synchronization.

The check error will abort the program. So I don't think the issue of `<<` really matters.